cartographer/io/serialization_format_migration.cc

```diff
@@ -0,0 +1,153 @@
+  }
+  if (legacy_data.has_trajectory_data()) {
+    auto& output_vector =
+        (*proto_map)[SerializedData::kTrajectoryDataFieldNumber];
+    output_vector.push_back({});
+    *output_vector.back().mutable_trajectory_data() =
+        legacy_data.trajectory_data();
+  }
+  if (legacy_data.has_imu_data()) {
+    auto& output_vector = (*proto_map)[SerializedData::kImuDataFieldNumber];
+    output_vector.push_back({});
+    *output_vector.back().mutable_imu_data() = legacy_data.imu_data();
+  }
+  if (legacy_data.has_odometry_data()) {
+    auto& output_vector = (*proto_map)[SerializedData::kOdometryData];
+    output_vector.push_back({});
+    *output_vector.back().mutable_odometry_data() = legacy_data.odometry_data();
+  }
+  if (legacy_data.has_fixed_frame_pose_data()) {
+    auto& output_vector =
+        (*proto_map)[SerializedData::kFixedFramePoseDataFieldNumber];
+    output_vector.push_back({});
+    *output_vector.back().mutable_fixed_frame_pose_data() =
+        legacy_data.fixed_frame_pose_data();
+  }
+  if (legacy_data.has_landmark_data()) {
+    auto& output_vector =
+        (*proto_map)[SerializedData::kLandmarkDataFieldNumber];
+    output_vector.push_back({});
+    *output_vector.back().mutable_landmark_data() = legacy_data.landmark_data();
+  }
+  return true;
+}
+
+ProtoMap ParseLegacyData(
+    cartographer::io::ProtoStreamReaderInterface* const input) {
+  ProtoMap proto_map;
+  CHECK(ReadPoseGraph(input, &proto_map))
+      << "Input stream seems to differ from original stream format. Could "
+         "not "
+         "read PoseGraph as first message.";
+  CHECK(ReadBuilderOptions(input, &proto_map))
+      << "Input stream seems to differ from original stream format. Could "
+         "not "
+         "read AllTrajectoryBuilderOptions as second message.";
+  while (DeserializeNext(input, &proto_map)) {
+  }
+  return proto_map;
+}
+
+mapping::proto::SerializationHeader CreateSerializationHeader() {
+  constexpr uint32_t kVersion1 = 1;
+  mapping::proto::SerializationHeader header;
+  header.set_format_version(kVersion1);
+  return header;
+}
+
+void SerializeToVersion1Format(
+    const ProtoMap& deserialized_data,
+    cartographer::io::ProtoStreamWriterInterface* const output) {
+  static const std::vector<int> kFieldSerializationOrder = {
+      SerializedData::kPoseGraphFieldNumber,
+      SerializedData::kAllTrajectoryBuilderOptionsFieldNumber,
+      SerializedData::kSubmapFieldNumber,
+      SerializedData::kNodeFieldNumber,
+      SerializedData::kTrajectoryDataFieldNumber,
+      SerializedData::kImuDataFieldNumber,
+      SerializedData::kOdometryDataFieldNumber,
+      SerializedData::kFixedFramePoseDataFieldNumber,
+      SerializedData::kLandmarkDataFieldNumber};
+
+  output->WriteProto(CreateSerializationHeader());
+  for (auto field_index : kFieldSerializationOrder) {
+    const auto proto_vector_it = deserialized_data.find(field_index);
+    if (proto_vector_it == deserialized_data.end()) continue;
```

Comment:
Is it needed to check whether `deserialized_data` has `field_index`? I would replace these three lines with:
` for (const auto& proto : deserialized_data[field_index]) {`

for range loops, `deserialized_data[]` will be evaluated only once.

Reply:
I could remove the check.
But you can only use the bracket operator on non-const maps since it would insert-if-not present. So I would vote for keeping all three lines here.

Reply:
yeah, right, did not think about "non-constness". Then leave it like that.